src/WebJobs.Script.WebHost/Metrics/FlexConsumptionMetricsPublisher.cs

```diff
@@ -0,0 +1,290 @@
+                };
+
+                FunctionExecutionTimeMS = FunctionExecutionCount = 0;
+                _instanceActivityStopwatch = ValueStopwatch.StartNew();
+            }
+
+            await PublishMetricsAsync(metrics);
+        }
+
+        private async void OnFunctionMetricsPublishTimer(object state)
+        {
+            await OnPublishMetrics();
+        }
+
+        private async Task PublishMetricsAsync(Metrics metrics)
+        {
+            try
+            {
+                bool metricsPublishEnabled = !string.IsNullOrEmpty(MetricsFilePath);
+                if (metricsPublishEnabled && !PrepareDirectoryForFile(_fileSystem))
+                {
+                    return;
+                }
+
+                string metricsContent = JsonConvert.SerializeObject(metrics);
+                _logger.PublishingMetrics(metricsContent);
+
+                if (metricsPublishEnabled)
+                {
+                    string fileName = $"{Guid.NewGuid().ToString().ToLower()}.json";
+                    string filePath = Path.Combine(MetricsFilePath, fileName);
+
+                    using (var streamWriter = _fileSystem.File.CreateText(filePath))
+                    {
+                        await streamWriter.WriteAsync(metricsContent);
+                    }
+                }
+            }
+            catch (Exception ex) when (!ex.IsFatal())
+            {
+                _logger.LogError(ex, "Error writing metrics file.");
+            }
+        }
+
+        private bool PrepareDirectoryForFile(IFileSystem fileSystem)
+        {
+            if (string.IsNullOrEmpty(MetricsFilePath))
+            {
+                return false;
+            }
+
+            // ensure the directory exists
+            _fileSystem.Directory.CreateDirectory(MetricsFilePath);
+
+            var metricsDirectoryInfo = _fileSystem.DirectoryInfo.FromDirectoryName(MetricsFilePath);
+            var files = metricsDirectoryInfo.GetFiles().OrderBy(p => p.CreationTime).ToList();
+
+            // ensure we're under the max file count
+            if (files.Count < _optionsMonitor.CurrentValue.MaxFileCount)
+            {
+                return true;
+            }
+
+            // we're at or over limit
+            // delete enough files that we have space to write a new one
+            int numToDelete = files.Count - _optionsMonitor.CurrentValue.MaxFileCount + 1;
+            var filesToDelete = files.Take(numToDelete).ToArray();
+
+            _logger.LogDebug($"Deleting {filesToDelete.Length} metrics file(s).");
+
+            bool success = true;
+            foreach (var file in filesToDelete)
+            {
+                try
+                {
+                    file.Delete();
+                }
+                catch (Exception ex) when (!ex.IsFatal())
+                {
+                    // best effort
+                    _logger.LogError(ex, "Error deleting metrics file.");
+                    success = false;
```

Comment:
Would there be a use case where we needed to delete 4 files and first 3 were successful and last one fails. So we end up returning `false` even though we deleted some files ( so we have space for new ones).

Instead of returning success in 192, should we execute the same code from 162 -168 ?

Reply:
Fixed